README
I kept my same dungeon generator code from the previous project except i added the professors object description class to load objects into the game. I made some pretty large changes to the game starting with the first level of the dungeon being a village with no monsters. Objects still spawn in the village. You spawn on stairs so you can instantly go into the dungeon where the monsters that were loaded from the file reside. I made changes to how the dungeon is displayed. Now you can see one layer into the rocks and they are printed as '#' now. The paths are then printed as '+'. To go up a level from the village you have to go all the way to the top border and find a '<'. Walk through it and you will enter a forest that is populated with monsters and objects. Another change is when you're in Look-Mode and you move over a monster or an object, you can hit SHIFT + I to get its information. NOTE: when you enter a new level, the previous level is saved into HOME/.rlg327/dungeon followed by the number of the dungeonlevel. you may want to delete these files after running.


CHANGELOG
	1/23/2017
	I started by creating dungeons.c which included a main() as well as a method generateRock() which will generate a dungeon filled completely with rocks of varying hardness except for the edges which are all the same hardness

	1/24/2017
	I was able to finish the random room generating function but i had a bug in the path generating function that lead to a ***stack smashing error***.

	1/25/2017
	I found the bug in the path generating function. It was due to a trying to access a field that was outside the array causing the program to crash

	1/28/2017
	Began the second project for assignment 1 continuing to use my same code from the previous assignment. Created 2 functions for reading and writing the binary file and well as the structure in main() to be able to handle load and save switches.

	1/30/2017
	I was able to implement code that was able to read and write files that I specifically created but there were bugs when i would try to read the test dungeons such as entire rooms being converted to paths after a save.

 	2/1/2017
	Discover the bug that prevented me from reading other files correctly. It was due to my code being unable to handle different numbers of rooms than what was specifically defined in my code.

	2/5/2017
	Started assignment 1.03 by making a change to my Makefile so I could work on new projects in seperate .c files and then link them together after compiling. I also began implementing dijkstra's algorithm by using Professor Sheaffer's heap.c, heap.h, and macros.h.

	2/7/2017
	First attempt at dijkstra's had many compile errors so I scrapped the whole thing and began looking at Professor Sheaffer's implementation of dijkstra for help.

	2/12/2017
	Completed an algorithm that could compile but when it ran it produced segmentation faults.

	2/13/2017
	Finished the algorithm for tunneling monsters and easily modified it generate a distance map for non-tunneling monsters as well.

	2/15/2017
	Modified dijkstra for a third time to create better dungeon paths than my old paths and added tags "--bunker", "--tame", and "--crazy" to generate dungeons with different path types (see README for more details).

	2/19/2017
	Started assignment 1.04 by creating a new file monsters.c and adding it to the makefile. I also added unimplemented functions to get an idea of the structure.

	2/20/2017
	Created structs for monsters, player characters, and events as well as initialized them and added them to the heap.

	2/21/2017
	Implemented movement functions for all of the 16 types of monsters as well as random movement for the player character.

	2/22/2017
	completed the assignment and debugged the various movement functions.

	2/27/2017
	Started the assigment and created an ncurses window that prints centered on the player character

	3/1/2017
	Implemented the different key presses so the player can move

	3/3/2017
	Added staircases that make a new random dungeon or load one that you've already visited

	3/18/2017
	Started assignment 1.06 by creating C++ classes and class headers for Player and Monster

	3/20/2017
	Implemented constructors and methods for the classes

	3/22/2017
	created wrapper functions that allowed an interface between C and C++ and went through and fixed all compile errors in the C part of code

	3/23/2017
	Finished project by debugging and implemented for of war that allows you to see dungeon that you've already visited and lets you see monster only in a 5 block radius.

	3/28/2017
	started project 1.07 by converting all my code over to c++ and fixing all the compile errors

	3/29/2017
	created a new monster description class and a dice class and began writing a function that can read a text file and create monster descriptions and print them out

	3/30/2017
	finished project 1.07 by debugging.

	4/2/2017
	Started project by loading monster into the dungeon and debugging

	4/3/2017
	Used the professor's code for object descriptions and created object class

	4/4/2017
	loaded objects into the dungeon

	4/5/2017
	added a village level on the first dungeon level and a forest level and debugged to finish project







core.h
//libraries
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <endian.h>
#include <limits.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <errno.h>
#include <ncurses.h>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <cstdlib>
#include <cstdio>
#include <cstring>

//namespaces
using namespace std;

//DUNGEON CONSTANTS
#define WIDTH 160
#define HEIGHT 105
#define HARDEST 255
#define ONE_HARD 85
#define ROOM '.'
#define PATH '+'
#define ROCK ' '
#define numRooms 25
#define EDGE 'E'
#define PC '@'
#define PC_SPEED 25
#define MAX_SPEED 20
#define MIN_SPEED 5
#define NUM_M_TYPES 16
#define NCURSES_WIDTH 80
#define NCURSES_HEIGHT 21
#define UP_STAIRCASE '<'
#define DOWN_STAIRCASE '>'
#define STD_MONS 15
#define EASY_MONS 5
#define HARD_MONS 25
#define EXP_MONS 50
#define DUNGEON_SIGHT_RADIUS 5
#define FOREST_SIGHT_RADIUS 5
#define MON_SIGHT_RADIUS 8
#define ROCK_CHAR '#'
#define WALL_CHAR '#'
#define TREE_CHAR '&'
#define GRASS_CHAR ';'
#define CARRY_SLOTS 10
#define SCROLL_SLOTS 10
#define SHOP_SLOTS 10
#define CONSUMABLE_SLOTS 20
#define STACK_CHAR '*'
#define LEVEL_SCALAR 8
#define OBJECT_SCALAR 10
#define VIT_SCALAR 10
#define TREE_HARD 10
#define DUNG_ENTRANCE 11
#define VERT_WALL 1
#define TOP_L_CORNER 2
#define TOP_R_CORNER 3
#define BOT_L_CORNER 4
#define BOT_R_CORNER 5
#define HORZ_WALL 6
#define FLOOR_OBJS 0
#define MON_DROP 1
#define DROP_RATE 5
#define BOSS_EXP_MULTIPLIER 2
#define BOSS_TELE 4
#define FROZEN_LENGTH 30
#define FIRST_BOSS_LEVEL -3
#define SECOND_BOSS_LEVEL -6

//KEY CONSTANTS
#define KEY_7 55
#define KEY_8 56
#define KEY_9 57
#define KEY_4 52
#define KEY_5 53
#define KEY_6 54
#define KEY_0 48
#define KEY_1 49
#define KEY_2 50
#define KEY_3 51
#define KEY_y 121
#define KEY_k 107
#define KEY_u 117
#define KEY_l 108
#define KEY_n 110
#define KEY_j 106
#define KEY_b 98
#define KEY_h 104
#define KEY_SPACE 32
#define KEY_L 76
#define KEY_Q 81
#define KEY_DOWNSTAIRS 62
#define KEY_UPSTAIRS 60
#define KEY_ESC 27
#define KEY_I 73
#define KEY_S1 70
#define KEY_S2 66
#define KEY_S3 126
#define KEY_S4 68
#define KEY_S5 69
#define KEY_S6 67
#define KEY_S7 72
#define KEY_S8 65
#define KEY_S9 126
#define KEY_a 97
#define KEY_b 98
#define KEY_c 99
#define KEY_d 100
#define KEY_e 101
#define KEY_f 102
#define KEY_g 103
#define KEY_h 104
#define KEY_i 105
#define KEY_w 119
#define KEY_t 116
#define KEY_x 120
#define KEY_m 109
#define KEY_o 111
#define KEY_p 112
#define KEY_q 113
#define KEY_r 114
#define KEY_s 115

//colors
#define COLOR_GRAY 8
#define COLOR_ROCK_DARK 9
#define COLOR_ROCK_ILUM 10
#define COLOR_DARK_GREEN 11
#define COLOR_RED_LAVA 12
#define COLOR_YELLOW_LAVA 13

//global variables
extern int roomDimensions[100][4];
extern int roomNumber;
extern char roomAndPath[WIDTH][HEIGHT];
extern unsigned char hardness[WIDTH][HEIGHT];
extern char characterMap[WIDTH][HEIGHT];
extern char objectMap[WIDTH][HEIGHT];
extern char staircases[WIDTH][HEIGHT];
extern unsigned char seen[WIDTH][HEIGHT];
extern int pcX;
extern int pcY;
extern char prevSpace;
extern int all_distances[WIDTH][HEIGHT];
extern int rp_distances[WIDTH][HEIGHT];
extern int dungeonLevel;
extern int highestReached;
extern int lowestReached;
extern int nextDungeon;
extern stringstream dungeonFile;
extern int killed;
extern int monsleft;
extern int difficulty;
extern int diffType[4];
extern int quit;
extern int numObjs;
extern int firstBossLevel;
extern int secondBossLevel;
extern int grassPattern[WIDTH][HEIGHT];
extern char grassTypes[4];
extern string villageName;
extern string villageNames[13];
extern bool isFrozen;
extern int frozenTurn;

//Dungeon enums
enum visible{
  HIDDEN,
  VISIBLE,
  SEEN
};

enum difficulty{
  easy,
  STD,
  hard,
  exp
};

typedef enum life_status {
  ALIVE,
  DEAD
} life_status_t;

enum color_pairs{
  blank,
  ILUM,
  DARK,
  R_DARK,
  MONST,
  R_ILUM,
  PC_COLOR,
  GRAS,
  D_GRAS,
  YEL,
  CY,
  BL,
  MAG,
  LAV_ROCK,
  LAV,
  Y_LAV
};

//file headers
#include "heap.h"
#include "macros.h"
#include "io_main.h"
#include "dijkstra.h"
#include "dice.h"
#include "object.h"
#include "object_description.h"
#include "monster.h"
#include "player.h"
#include "game.h"
#include "dice.h"
#include "monster_desc.h"

extern player* main_character;
extern monster *mondun[WIDTH][HEIGHT];
extern object *objdun[WIDTH][HEIGHT];
extern monster *boss;



